pkg/disttask/framework/proto/node.go

```diff
@@ -23,3 +23,44 @@ type ManagedNode struct {
 	Role     string
 	CPUCount int
 }
+
+// NodeResource is the resource of the node.
+// exported for test.
+type NodeResource struct {
+	TotalCPU  int
+	TotalMem  int64
+	TotalDisk uint64
+}
+
+// NewNodeResource creates a new NodeResource.
+func NewNodeResource(totalCPU int, totalMem int64) *NodeResource {
+	return &NodeResource{
+		TotalCPU:  totalCPU,
+		TotalMem:  totalMem,
+		TotalDisk: 0,
+	}
+}
+
+// NewNodeResourceWithDisk creates a new NodeResource with totalDisk initialized.
+func NewNodeResourceWithDisk(totalCPU int, totalMem int64, totalDisk uint64) *NodeResource {
+	return &NodeResource{
+		TotalCPU:  totalCPU,
+		TotalMem:  totalMem,
+		TotalDisk: totalDisk,
+	}
+}
+
+// GetStepResource gets the step resource according to concurrency.
+func (nr *NodeResource) GetStepResource(concurrency int) *StepResource {
+	return &StepResource{
+		CPU: NewAllocatable(int64(concurrency)),
+		// same proportion as CPU
+		Mem: NewAllocatable(int64(float64(concurrency) / float64(nr.TotalCPU) * float64(nr.TotalMem))),
+	}
+}
+
+// GetTaskDiskResource gets available disk for a task.
+func (nr *NodeResource) GetTaskDiskResource(concurrency int, quotaHint uint64) uint64 {
```

Comment:
is `quotaHint` used to allocate less resource for some task? why do we need it if we cannot allow some task to alloc more resource?

and it should be part of `StepResource`, no need a special API

Reply:
`StepResource` is only used by the step executor. We need to know the resource in the scheduler before splitting subtasks.

Reply:
> is quotaHint used to allocate less resource for some task?

Yes.

> why do we need it if we cannot allow some task to alloc more resource?

Because other factors may affect the resource allocation, for example, `@@tidb_ddl_disk_quota` can change dynamically.

Reply:
if the resource per slot for a subtask is always <= `the mount of disk per slot` = `D`, seems ok to just use `D`, no need to adjust

Reply:
We can't determine which one is larger, `tidb_ddl_disk_quota` or actual disk capacity. For example, a user has a 2TB node, but only wants to use 500GB for DXF, then the user can set `tidb_ddl_disk_quota` to 500GB

Reply:
no need to consider `tidb_ddl_disk_quota`, we can always use the `total_disk` we got by inspecting disk stat.